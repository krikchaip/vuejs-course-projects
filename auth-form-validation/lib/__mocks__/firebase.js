export const emailPasswordSignUp = jest.fn().mockResolvedValue({
  kind: 'identitytoolkit#SignupNewUserResponse',
  idToken: 'SECRET_ID_TOKEN',
  email: 'test@test.com',
  refreshToken: 'SECRET_REFRESH_ID_TOKEN',
  expiresIn: '3600',
  localId: 'p0nYsW@gG3R'
})

export const emailPasswordSignIn = jest.fn().mockResolvedValue({
  kind: 'identitytoolkit#SignupNewUserResponse',
  idToken: 'SECRET_ID_TOKEN',
  email: 'test@test.com',
  refreshToken: 'SECRET_REFRESH_ID_TOKEN',
  expiresIn: '3600',
  localId: 'p0nYsW@gG3R',
  registered: true
})

export const addUserRecord = jest.fn((data, _) => {
  delete data.UID
  return Promise.resolve(data)
})

export const getUserRecord = jest.fn().mockResolvedValue({
  email: 'test@test.com',
  age: '23',
  country: 'USA'
})

export default {
  emailPasswordSignUp,
  emailPasswordSignIn,
  addUserRecord,
  getUserRecord
}
