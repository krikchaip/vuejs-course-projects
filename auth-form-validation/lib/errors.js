class CustomError extends Error {
  constructor(message) {
    super(message);
    this.name = this.constructor.name;
  }
}

class FirebaseError extends CustomError {}
export class APIKeyError extends FirebaseError {}
export class TargetDBError extends FirebaseError {}
