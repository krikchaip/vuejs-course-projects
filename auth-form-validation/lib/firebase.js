import _ from 'ramda'
import axios, { AxiosResponse } from 'axios'
import { APIKeyError, TargetDBError } from 'lib/errors'

axios.interceptors.response.use(
  value => value.data,
  error => error.response.data
)

const BASE_ENDPOINT = 'https://www.googleapis.com/identitytoolkit/v3/relyingparty/'

/**
 * @param {String} type Endpoint type
 * @param {String} API_KEY Firebase Web API Key
 */
function toEndpoint(type, API_KEY) {
  return BASE_ENDPOINT + type + '?key=' + API_KEY
}

/**
 * @param {String} uid Firebase user UID
 * @param {String} ID_TOKEN Authenticate request token
 */
function toRecordURL(uid, ID_TOKEN) {
  return getFirebaseDBUrl() + '/users/' + uid + '.json' + '?auth=' + ID_TOKEN
}

/**
 * @param {String} key to look up in process.env
 */
function IOgetEnv(key) {
  /** @type {String | undefined} */
  const value = _.prop(key, process.env)

  if(value) { return value }
  else { throw new ReferenceError(`process.env.${key} not found`) }
}

/**
 * @returns {String} FIREBASE_API_KEY
 */
function getFirebaseAPIKey() {
  try { return IOgetEnv('VUE_APP_FIREBASE_API_KEY') }
  catch(_) { throw new APIKeyError('FIREBASE_API_KEY not found') }
}

/**
* @returns {String} FIREBASE_DB_URL
*/
function getFirebaseDBUrl() {
  try { return IOgetEnv('VUE_APP_FIREBASE_DB_URL') }
  catch(_) { throw new TargetDBError('FIREBASE_DB_URL not found') }
}

/**
 * see firebase [documentation](https://goo.gl/fszB6r) for return value
 * @param {String} email
 * @param {String} password
 */
export async function emailPasswordSignUp(email, password) {
  return axios.post(
    toEndpoint('signupNewUser', getFirebaseAPIKey()),
    { email, password, returnSecureToken: true }
  )
}

/**
* see firebase [documentation](https://goo.gl/DHfXKw) for return value
* @param {String} email
* @param {String} password
*/
export async function emailPasswordSignIn(email, password) {
  return axios.post(
    toEndpoint('verifyPassword', getFirebaseAPIKey()),
    { email, password, returnSecureToken: true }
  )
}

/**
 * @typedef UserData
 * @type {{ [a: string]: any, UID: String, email: String }}
 */

/**
 * @param {UserData} userData
 * @param {String} ID_TOKEN Authenticate request token
 * @return {Promise<AxiosResponse<UserData>>} the data we wrote to the database
 */
export async function addUserRecord(userData, ID_TOKEN) {
  return axios.put(
    toRecordURL(userData.UID, ID_TOKEN),
    { ..._.omit(['UID'], userData) }
  )
}

export default {
  emailPasswordSignUp,
  emailPasswordSignIn,
  addUserRecord
}
