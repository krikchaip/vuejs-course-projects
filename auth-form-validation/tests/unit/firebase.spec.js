import axios from 'axios'
import { APIKeyError, TargetDBError } from 'lib/errors'
import firebase from 'lib/firebase'

jest.mock('axios')

const API_KEY = process.env.VUE_APP_FIREBASE_API_KEY
const DB_URL = process.env.VUE_APP_FIREBASE_DB_URL

afterEach(() => {
  jest.resetAllMocks()
  process.env.VUE_APP_FIREBASE_API_KEY = API_KEY
  process.env.VUE_APP_FIREBASE_DB_URL = DB_URL
})

describe('emailPasswordSignUp', () => {
  it('reject APIKeyError when no API_KEY provided', async () => {
    delete process.env.VUE_APP_FIREBASE_API_KEY
    await expect(firebase.emailPasswordSignUp('winner', '123456'))
      .rejects.toBeInstanceOf(APIKeyError)
  })

  describe('with API_KEY', () => {
    beforeEach(async () => {
      process.env.VUE_APP_FIREBASE_API_KEY = 'something_secret'
      await firebase.emailPasswordSignUp('winner', '123456')
    })

    it('make post request to the API endpoint', () => {
      // @ts-ignore
      expect(axios.post.mock.calls[0][0]).toMatchSnapshot()
    })

    it('payload data contain email, password and returnSecureToken', () => {
      // @ts-ignore
      expect(axios.post.mock.calls[0][1]).toMatchObject({
        email: 'winner',
        password: '123456',
        returnSecureToken: true
      })
    })
  })
})

describe('emailPasswordSignIn', () => {
  it('reject APIKeyError when no API_KEY provided', async () => {
    delete process.env.VUE_APP_FIREBASE_API_KEY
    await expect(firebase.emailPasswordSignIn('winner', '123456'))
      .rejects.toBeInstanceOf(APIKeyError)
  })

  describe('with API_KEY', () => {
    beforeEach(async () => {
      process.env.VUE_APP_FIREBASE_API_KEY = 'something_secret'
      await firebase.emailPasswordSignIn('winner', '123456')
    })

    it('make post request to the API endpoint', () => {
      // @ts-ignore
      expect(axios.post.mock.calls[0][0]).toMatchSnapshot()
    })

    it('payload data contain email, password and returnSecureToken', () => {
      // @ts-ignore
      expect(axios.post.mock.calls[0][1]).toMatchObject({
        email: 'winner',
        password: '123456',
        returnSecureToken: true
      })
    })
  })
})

describe('addUserRecord', () => {
  const userData = {
    UID: 'SeCrEt1234',
    email: 'test@test.com',
  }
  const ID_TOKEN = 'auth_token_for_realtime_db'

  it('reject TargetDBError when no DB_URL provided', async () => {
    delete process.env.VUE_APP_FIREBASE_DB_URL
    await expect(firebase.addUserRecord(userData, ID_TOKEN))
      .rejects.toBeInstanceOf(TargetDBError)
  })

  describe('with DB_URL', () => {
    beforeEach(async () => {
      process.env.VUE_APP_FIREBASE_DB_URL = '[URL_TO_FIREBASE_REALTIME_DB]'
      await firebase.addUserRecord(userData, ID_TOKEN)
    })

    it('make put request to /users/[UID].json?auth=[ID_TOKEN]', () => {
      // @ts-ignore
      expect(axios.put.mock.calls[0][0]).toMatchSnapshot()
    })

    it('no UID field present in payload data', () => {
      // @ts-ignore
      expect(axios.put.mock.calls[0][1]).toEqual({ email: 'test@test.com' })
    })
  })
})
