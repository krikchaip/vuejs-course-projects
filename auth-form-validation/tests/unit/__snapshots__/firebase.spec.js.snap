// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addUserRecord with DB_URL make put request to /users/[UID].json?auth=[ID_TOKEN] 1`] = `"[URL_TO_FIREBASE_REALTIME_DB]/users/SeCrEt1234.json?auth=auth_token_for_realtime_db"`;

exports[`emailPasswordSignIn with API_KEY make post request to the API endpoint 1`] = `"https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyPassword?key=something_secret"`;

exports[`emailPasswordSignUp with API_KEY make post request to the API endpoint 1`] = `"https://www.googleapis.com/identitytoolkit/v3/relyingparty/signupNewUser?key=something_secret"`;
